### PR TITLE
Adding support for multi statement queries

### LINKED ISSUE #8: Handle multi-statement queries
We need to be able to handle queries like this:
```ts
const query = `
    CREATE TABLE ids(id integer); 
    INSERT INTO ids(id) values(1); 
    INSERT INTO ids(id) values(2);
`;

await client.query(query);
```

I'm not sure how to nicely handle results, currently `QueryResult` has `rows` property that stores data returned from query. One idea is to subclass `QueryResult` and create `MultiQueryResult` that has slight different API and leave it up to user to handle that properly. 

Comments welcome

### PR DESCRIPTION
**Pull request by halvardssm**
fixes #8

### CODE REVIEW

File: client.ts [hunk 1]
```diff
@@ -24,6 +25,21 @@ export class Client {
     return await this._connection.query(query);
   }
 
+  async transaction(
```
Comment (bartlomieju) [inline client.ts:28]: This method doesn't really have anything in common with `transaction`... Maybe just `multiQuery`?

File: client.ts [hunk 2]
```diff
@@ -24,6 +25,21 @@ export class Client {
...
+    text: string | string[],
+    args?: unknown[],
+    name?: string,
```
Comment (bartlomieju) [inline client.ts:31]: This argument is unused

File: client.ts [hunk 3]
```diff
@@ -24,6 +25,21 @@ export class Client {
...
+    encoder?: (arg: unknown) => EncodedArg,
```
Comment (bartlomieju) [inline client.ts:32]: Ditto

File: client.ts (hunk 1 above)
Comment (halvardssm) [reply client.ts]: sure, fixed now

File: client.ts (hunk 2 above)
Comment (halvardssm) [reply client.ts]: yeah, I used the `QueryConfig` interface for the arguments. I added a new commit where I removed it, let me know if this is preffered

File: client.ts (hunk 3 above)
Comment (halvardssm) [reply client.ts]: same as above

File: client.ts [hunk 4]
```diff
@@ -17,13 +17,28 @@ export class Client {
...
-    const query = new Query(text, ...args);
+    const query = new Query({ text, args });
     return await this._connection.query(query);
   }
 
+  async multiQuery(
+    text: string | string[],
+    args?: unknown[],
+  ): Promise<QueryResult[]> {
+    if (!Array.isArray(text)) text = text.split(";").map((el) => el.trim());
```
Comment (bartlomieju) [inline client.ts:31]: This line seems brittle; can you add some tests showcasing that it works as intended?

Comment (halvardssm) [reply client.ts:31]: hehe, yeah, currently there are a lot of tests failing, will fix all the tests when we have decided on the final implementation 💪

**Comment by halvardssm** [comment]
I see that name and encoder is used by Query (query.ts), should i also remove this if it is not used?

**Comment by halvardssm** [comment]
> I see that name and encoder is used by Query (query.ts), should i also remove this if it is not used?

@bartlomieju ^

**Comment by bartlomieju** [comment]
> > I see that name and encoder is used by Query (query.ts), should i also remove this if it is not used?
> 
> @bartlomieju ^

Yes, sounds good to me

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
+name: ci
+
+on: [push, pull_request, release]
```
Comment (bartlomieju) [inline .github/workflows/ci.yml:3]: @halvardssm please open a separate PR for CI update - try to keep PRs focused on single problem

Comment (halvardssm) [reply .github/workflows/ci.yml:3]: @bartlomieju Sure, added one here #109

**Comment by halvardssm** [comment]
@bartlomieju It is merged and ready now 👍

**Comment by halvardssm** [comment]
@bartlomieju Quick question, should I resolve the conversations? Also, is there anything else we should include here, or can we merge it? 😄

File: tests/data_types.ts [hunk 6]
```diff
@@ -20,11 +20,11 @@ testClient(async function inet() {
   const inet = "127.0.0.1";
   const insertRes = await CLIENT.query(
     "INSERT INTO data_types (inet_t) VALUES($1)",
-    inet,
+    { args: [inet] },
```
Comment (bartlomieju) [inline tests/data_types.ts:23]: What's the reason for this change?

File: tests/queries.ts [hunk 7]
```diff
@@ -42,5 +48,95 @@ testClient(async function binaryType() {
...
+  assertEquals(
+    result[1].rows[0][0].toUTCString(),
+    new Date(expectedDate).toUTCString(),
+  );
+
+  assertEquals(result[2].rows.length, 2);
+
+  await CLIENT.multiQuery([{
+    text: "INSERT INTO bytes VALUES($1);",
+    config: { args: [expectedBytes] },
```
Comment (bartlomieju) [inline tests/queries.ts:94]: Why is `config` needed? I think `{ text, args }` interface was sufficient for this method

File: tests/data_types.ts (hunk 6 above)
Comment (halvardssm) [reply tests/data_types.ts:23]: No worries, with the release of 1.0 right around the corner I completely understand it :) 
I might have misunderstood what you mean then. I the multiquery added support for splitting the queries if you just pass in a big query string, or if you pass in an array of strings. QueryType is also a part of the signature, so if you have muliple queries with independent arguments, that is also supported. This is currently the signature:

```
query(
    text: string,
    config?: QueryConfig,
  ): Promise<QueryResult>;

async multiQuery(
    queries: string | string[] | QueryType[],
  ): Promise<QueryResult[]>;

export type QueryType = {
  text: string;
  config?: QueryConfig;
};

export interface QueryConfig {
  args?: Array<unknown>;
  name?: string;
  encoder?: (arg: unknown) => EncodedArg;
}
```

The option would be to have this:
```
query(
    text: string,
    args?: Array<unknown>,
  ): Promise<QueryResult>;

async multiQuery(queries: QueryType[]): Promise<QueryResult[]>;

export type QueryType = {
  text: string;
  args?: Array<unknown>;
};
```

Is the latter maybe better to reduce the logic and/or possible bugs?

File: tests/queries.ts (hunk 7 above)
Comment (halvardssm) [reply tests/queries.ts:94]: I might have misunderstood what you meant with removing `encoder` and `name`, I thought you mean from only these functions, should I prune all the methods from this (or the whole module even)?

File: tests/data_types.ts (hunk 6 above)
Comment (bartlomieju) [reply tests/data_types.ts:23]: The second one looks exactly like current `QueryConfig` interface :) so I guess no need to change the type?

File: tests/queries.ts (hunk 7 above)
Comment (bartlomieju) [reply tests/queries.ts:94]: No, I mean that I don't understand why `QueryConfig { text, args }` was replaced with `QueryType { text, config: { args } }` - it seems they essentially do the same thing - if so then probably no need to change `QueryConfig` at all

Comment (halvardssm) [reply tests/queries.ts:94]: I thought it might be cleaner to have a config input where all the settings/args etc. can be inside so that it is future proof for future settings/options. Do you think it's better to keep it as it is?

Comment (bartlomieju) [reply tests/queries.ts:94]: Yeah I think so - `QueryConfig` can still be extended in the future with additional fields

Comment (halvardssm) [reply tests/queries.ts:94]: ok, then i will revert it

File: tests/data_types.ts (hunk 6 above)
Comment (halvardssm) [reply tests/data_types.ts]: fair enough :)

File: tests/queries.ts (hunk 7 above)
Comment (halvardssm) [reply tests/queries.ts]: @bartlomieju It should be all good now 👍
